cmd/fleetctl/desktop.go

```diff
@@ -0,0 +1,180 @@
+package main
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+
+	"github.com/fleetdm/fleet/v4/orbit/pkg/constant"
+	"github.com/fleetdm/fleet/v4/pkg/secure"
+	"github.com/rs/zerolog"
+	zlog "github.com/rs/zerolog/log"
+	"github.com/urfave/cli/v2"
+)
+
+func desktop() *cli.Command {
+	return &cli.Command{
+		Name:        "desktop",
+		Aliases:     nil,
+		Usage:       "Creates the Fleet Desktop Application",
+		Description: "The desktop command allows the creation and packaging of the Fleet Desktop Application from source.",
+		Flags: []cli.Flag{
+			&cli.StringFlag{
+				Name:     "platform",
+				Usage:    "Target platform (currently supports \"macos\" only)",
+				Required: true,
+			},
+			&cli.BoolFlag{
+				Name:  "verbose",
+				Usage: "Log detailed information when building the application",
+			},
+		},
+		Action: func(c *cli.Context) error {
+			if c.String("platform") != "macos" {
+				return fmt.Errorf("unsupported platform: %s", c.String("platform"))
+			}
+
+			if !c.Bool("verbose") {
+				zlog.Logger = zerolog.Nop()
+			}
+
+			if err := createMacOSApp(); err != nil {
+				return err
+			}
+
+			return nil
+		},
+	}
+}
+
+const infoPList = `<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
+	<key>CFBundleExecutable</key>
+	<string>fleet-desktop</string>
+	<key>CFBundleIdentifier</key>
+	<string>com.fleetdm.desktop</string>
+	<key>CFBundleInfoDictionaryVersion</key>
+	<string>6.0</string>
+	<key>CFBundleName</key>
+	<string>fleet-desktop</string>
+	<key>CFBundlePackageType</key>
+	<string>APPL</string>
+	<key>CFBundleShortVersionString</key>
+	<string>0.0.1</string>
+	<key>CFBundleVersion</key>
+	<string>0.0.1</string>
+	<!-- avoid having a blurry icon and text -->
+	<key>NSHighResolutionCapable</key>
+	<string>True</string>
+	<!-- avoid showing the app on the Dock -->
+	<key>LSUIElement</key>
+	<string>1</string>
+</dict>
+</plist>
+`
+
+func createMacOSApp() error {
+	if runtime.GOOS != "darwin" {
+		return errors.New("The \"Fleet Desktop\" macOS app can only be created from macOS")
+	}
+
+	const appDir = "Fleet Desktop.app"
+	defer os.RemoveAll(appDir)
+
+	contentsDir := filepath.Join(appDir, "Contents")
+	macOSDir := filepath.Join(contentsDir, "MacOS")
+	for _, dir := range []string{macOSDir} {
+		if err := secure.MkdirAll(dir, constant.DefaultDirMode); err != nil {
+			return fmt.Errorf("create directories: %w", err)
+		}
+	}
+
+	infoFile := filepath.Join(contentsDir, "Info.plist")
+	ioutil.WriteFile(infoFile, []byte(infoPList), 0o644)
+
+	cmd := exec.Command("go", "build", "-o", filepath.Join(macOSDir, "fleet-desktop"), "./"+filepath.Join("orbit", "cmd", "desktop"))
```

Comment:
This is certainly ok at this stage, but the end result shouldn't need the fleetctl user to have go installed in their path. This is probably stupidly obvious, but I felt compelled to mention it.

Reply:
This command (`fleetctl desktop`) is to be used only by TUF maintainers. What other alternative can you think of?

Reply:
Github Action with official releases? Wouldn't this be used as part of `fleetctl package` by users? In that case, it would work the same as with orbit, which I don't think we build there.

Reply:
> Github Action with official releases?

I see. We could just use `fleetctl desktop` command in Github CI to release.
Or, I could also move it to `tools` instead. (We will eventually need facilities for people managing their TUF servers.)

> Wouldn't this be used as part of fleetctl package by users?

- `fleetctl desktop` (or a tool under `tools`) would help build the target for desktop.
- `fleetctl updates add` is used to push such target to TUF.
- `fleetctl package` fetches targets from TUF.